demos/GradientDescent.cpp

```diff
@@ -0,0 +1,139 @@
+//--------------------------------------------------------------------*- C++ -*-
+// clad - The C++ Clang-based Automatic Differentiator
+//
+// Demo of a simple gradient decent algorithm to predict the feautres for a
+// liner relationship. Clad is used to calculate the gradient of the
+// cost function.
+//
+//----------------------------------------------------------------------------//
+
+// To run the demo please type:
+// path/to/clang  -Xclang -add-plugin -Xclang clad -Xclang -load -Xclang \
+// path/to/libclad.so  -I../include/ -x c++ -lstdc++ -lm GradientDescent.cpp
+//
+// A typical invocation would be:
+// ../../../../obj/Debug+Asserts/bin/clang  -Xclang -add-plugin -Xclang clad \
+// -Xclang -load -Xclang ../../../../obj/Debug+Asserts/lib/libclad.dylib     \
+// -I../include/ -x c++ -lstdc++ -lm GradientDescent.cpp
+//
+// To plot the results install gnuplot and type:
+// gnuplot -e "plot 'dataset_gd.dat' with points pt 7; replot 'out_gd.dat' using 1:2 with lines; pause -1"
+
+#include <iostream>
+#include <fstream>
+#include <cstdlib>
+#include <vector>
+#include <algorithm>
+
+// Necessary for clad to work include
+#include "clad/Differentiator/Differentiator.h"
+
+// The hypothesis function for the quadratic problem
+// theta_0, theta_1 are the features to learn and x is the input data
+double f(double theta_0, double theta_1, double x) {
+    return theta_0 + theta_1 * x;
+}
+
+// Structure to represent input dataset
+// x, y represents the input and output data respectively
+// size represents the number of samples
+// learning_rate represents the learning rate
+struct Dataset {
+
+    std::vector<double> x;
+    std::vector<double> y;
+    size_t size = 50;
+    double learning_rate = 1e-5;
+
+    // Populate the data set with random data
+    Dataset() {
+
+        // For plotting data later
+        std::ofstream out("dataset_gd.dat");
+
+        for(size_t i = 0; i < size; i++) {
+
+            x.push_back( (std::rand()%20) );
+            y.push_back( f(2 + std::rand()%25, 5, x[i]) );
```

Comment:
```suggestion
            y.push_back(f(2 + std::rand()%25, 5, x[i]));
```

Although we may need extra variables to describe the meaning of these values.